src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/Helper.cs

```diff
@@ -0,0 +1,247 @@
+        {
+            // The full root directory is the relevant special folder from Environment.GetFolderPath() plus "IsolatedStorage"
+            // and a set of random directory names if not roaming. The current identity gets tacked on after this.
+
+            if (IsRoaming(scope))
+            {
+                if (string.IsNullOrEmpty(s_RoamingUserRootDirectory))
+                    s_RoamingUserRootDirectory = GetDataDirectory(scope);
+                return s_RoamingUserRootDirectory;
+            }
+
+            if (IsMachine(scope))
+            {
+                if (string.IsNullOrEmpty(s_MachineRootDirectory))
+                    s_MachineRootDirectory = GetDataDirectory(scope);
+                return s_MachineRootDirectory;
+            }
+
+            if (string.IsNullOrEmpty(s_UserRootDirectory))
+                s_UserRootDirectory = GetDataDirectory(scope);
+
+            return s_UserRootDirectory;
+        }
+
+        private static string GetRandomDirectory(string rootDirectory, IsolatedStorageScope scope)
+        {
+            string randomDirectory = GetExistingRandomDirectory(rootDirectory);
+            if (string.IsNullOrEmpty(randomDirectory))
+            {
+                Mutex m = CreateMutexNotOwned(rootDirectory);
+                if (!m.WaitOne())
+                    throw new IsolatedStorageException(SR.IsolatedStorage_Init);
+
+                try
+                {
+                    randomDirectory = GetExistingRandomDirectory(rootDirectory);
+                    if (string.IsNullOrEmpty(randomDirectory))
+                    {
+                        // Someone else hasn't created the directory before we took the lock
+                        randomDirectory = Path.Combine(rootDirectory, Path.GetRandomFileName(), Path.GetRandomFileName());
+                        CreateDirectory(randomDirectory, scope);
+                    }
+                }
+                finally
+                {
+                    m.ReleaseMutex();
+                }
+            }
+
+            return randomDirectory;
+        }
+
+        private static string GetExistingRandomDirectory(string rootDirectory)
+        {
+            // Look for an existing random directory at the given root
+            // (a set of nested directories that were created via Path.GetRandomFileName())
+
+            // Older versions of the desktop framework created longer (24 character) random paths and would
+            // migrate them if they could not find the new style directory.
+            foreach (string directory in Directory.GetDirectories(rootDirectory))
+            {
+                if (Path.GetDirectoryName(directory)?.Length == 12)
+                {
+                    foreach (string subdirectory in Directory.GetDirectories(directory))
+                    {
+                        if (Path.GetDirectoryName(subdirectory)?.Length == 12)
```

Comment:
What causes directories to be created in the rootDirectory?  If other actors can create directories, what's to say their lengths aren't 12?  And if other actors can't create directories, why are we checking the length?